src/solve/slg/implementation/mod.rs

```diff
@@ -233,7 +233,176 @@ impl context::Environment<SlgContext> for Arc<Environment> {
     }
 }
 
-impl context::Substitution<SlgContext> for Substitution {}
+impl context::Substitution<SlgContext> for Substitution {
+    fn may_invalidate(&mut self, subst: &Substitution) -> bool {
+        self.parameters
+            .iter()
+            .zip(&subst.parameters)
+            .any(|(p1, p2)| MayInvalidate.aggregate_parameters(p1, p2))
+    }
+}
+
+// This is a struct in case we need to add state at any point like in AntiUnifier
+struct MayInvalidate;
+
+impl MayInvalidate {
+    fn aggregate_parameters(&mut self, p1: &Parameter, p2: &Parameter) -> bool {
+        match (p1, p2) {
+            (ParameterKind::Ty(ty1), ParameterKind::Ty(ty2)) => {
+                self.aggregate_tys(ty1, ty2)
+            }
+            (ParameterKind::Lifetime(l1), ParameterKind::Lifetime(l2)) => {
+                self.aggregate_lifetimes(l1, l2)
+            }
+            (ParameterKind::Ty(_), _) | (ParameterKind::Lifetime(_), _) => {
+                panic!("mismatched parameter kinds: p1={:?} p2={:?}", p1, p2)
+            }
+        }
+    }
+
+    // Returns true if the two types could be unequal.
+    fn aggregate_tys(&mut self, ty0: &Ty, ty1: &Ty) -> bool {
+        match (ty0, ty1) {
+            (Ty::Var(_), Ty::Var(_)) => false,
```

Comment:
A couple of things here. First, I think we need to thread the inference table for the new answer to this code -- and maybe we ought to thread the inference table for the current answer too, for uniformity (though I think the aggregator never unifies, so it's not strictly needed in that case). That way we can begin with a "normalizing" step to get rid of inference variables that have bound values.
Then we can start this function with a "shallow normalize" step like so:

```rust
if let Some(ty0) = self.infer0.normalize_shallow(ty0) {
    return self.aggregate_tys(ty0, ty1);
} else if let Some(ty1) = self.infer1.normalize_shallow(ty1) {
    return self.aggregate_tys(ty0, ty1);
}
```

just like what you see here in unification (and you will find elsewhere):

https://github.com/rust-lang-nursery/chalk/blob/d03d8f9e7bfb90d2fdfec86b5f15a7ce432547e5/src/solve/infer/unify.rs#L92-L95

Next, once we've done that, we can tweak this a bit further:

```rust
match (new, current) { // using the names I suggested above =)
    (_, Ty::Var(_)) => {
        // If the aggregate solution already has an inference variable here,
        // then no matter what type we produce, the aggregate cannot get 'more generalized'
        // than it already is. So return false, we cannot invalidate.
        false
    }

    (Ty::Var(_), _) => {
        // If we see a type variable in the potential future solution, we have to be conservative.
        // We don't know what type variable will wind up being! Remember that the future solution could
        // be any instantiation of `ty0` -- or it could leave this variable unbound, if the result is true for
        // all types.
        true
    }

    ...
}
```

Reply:
I have to think though about how to make a test that will cause the existing code to go wrong =)